topic/machine-learning/llm-langchain/cratedb-vectorstore-rag-openai.ipynb

```diff
@@ -311,8 +305,8 @@
     "# The embeddings will be stored in CrateDB using the FLOAT_VECTOR type.\n",
     "df = pd.DataFrame(list(zip(pages_text, pages_embeddings)),columns =['text', 'embedding'])\n",
     "\n",
-    "create_table = text(\"CREATE TABLE IF NOT EXISTS text_data (text TEXT, embedding FLOAT_VECTOR(1536))\")\n",
-    "engine = create_engine(CONNECTION_STRING, echo=False)\n",
+    "create_table = sa.text(\"CREATE TABLE IF NOT EXISTS text_data (text TEXT, embedding FLOAT_VECTOR(1536))\")\n",
+    "engine = sa.create_engine(CONNECTION_STRING, echo=False)\n",
```

Comment:
When adding software tests here, we are observing this error.
```python
E           ProgrammingError: (crate.client.exceptions.ProgrammingError) RelationAlreadyExists[Relation 'testdrive.text_data' already exists.]
E           [SQL:
E           CREATE TABLE text_data (
E           	text STRING,
E           	embedding STRING
E           )
E
E           ]
E           (Background on this error at: https://sqlalche.me/e/20/f405)
E           ProgrammingError: (crate.client.exceptions.ProgrammingError) RelationAlreadyExists[Relation 'testdrive.text_data' already exists.]
E           [SQL:
E           CREATE TABLE text_data (
E           	text STRING,
E           	embedding STRING
E           )
E
E           ]
E           (Background on this error at: https://sqlalche.me/e/20/f405)

../../../.venv/lib/python3.11/site-packages/nbclient/client.py:862: CellExecutionError
```

Reply:
832c3bc176a fixes it. Indeed, we apparently need a runtime fix here, when using a non-standard schema (`doc` vs. `testdrive`).

```
# TODO: Bring this into the `crate-python` driver.
from cratedb_toolkit.sqlalchemy.patch import patch_inspector
patch_inspector()
```